### ISSUE TITLE
Render single frame for video formats (e.g., gifv/mp4)

### ISSUE THREAD

**Issue by 9ary** [tachiyomiorg/tachiyomi#9149]
### Describe your suggested feature

We maintain some automatically generated Cubari sources over at https://github.com/catgirl-v/cubari. One of the sources (ADHDinos, scraped from reddit) contains an animated strip (chapter 70), and as far as I can tell, reddit will only serve video formats. Imgur does this too with its "gifv" (which is just mp4 with a different extension).

Video codecs have been replacing the aging animated GIF format over the past decade because they offer much better compression (higher quality at much lower file sizes). Because we run our scrapers on github actions, we can't really afford to transcode the files on our end.

Currently, Tachiyomi shows an error message and offers to open the "image" in webview. It would be nice to support this natively.

### Other details

I believe this is different from #2262 because the videos in question are supposed to behave exactly like animated GIFs (short loop, no player controls, no sound). I also don't expect loading from sources such as youtube, or support for more complex streaming-optimized containers (DASH, HLS).

See also https://github.com/subject-f/cubarimoe/issues/17.

### Acknowledgements

- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open or closed issue.
- [X] I have written a short but informative title.
- [X] If this is an issue with an extension, I should be opening an issue in the [extensions repository](https://github.com/tachiyomiorg/tachiyomi-extensions/issues/new/choose).
- [X] I have updated the app to version **[0.14.5](https://github.com/tachiyomiorg/tachiyomi/releases/latest)**.
- [X] I will fill out all of the requested information in this form.

**Comment by arkon** [comment]
This is realistically the same thing as #2262 since like you mentioned, `gifv` is really just a full-on `mp4` video. The lack of sound or UI controls doesn't really change much in terms of needing to actually support video. I don't think it'd make sense to have to add/maintain an entire video playing stack into the app just for a few rare non-standard animated "images".

We do already support GIF and animated WebP, which are "real" animated image formats. We could maybe support animated AVIF in the feature, if that ever takes off (which it probably won't since Chrome is removing it).

Note that you can run `ffmpeg` on GitHub Actions.

**Comment by 9ary** [comment]
> This is realistically the same thing as #2262 since like you mentioned, `gifv` is really just a full-on `mp4` video. The lack of sound or UI controls doesn't really change much in terms of needing to actually support video. I don't think it'd make sense to have to add/maintain an entire video playing stack into the app just for a few rare non-standard animated "images".

This is understandable, video codecs are usually part of different libraries/APIs so even with shortcuts, it would still be a non-trivial addition. It's still less complex than a complete video player though.

The Cubari devs have accepted this feature for their frontend, but obviously it's a lot less difficult for them to implement since the browser does all the heavy lifting.

> We do already support GIF and animated WebP, which are "real" animated image formats.

WebP (both still and animated) is much rarer in the wild than MP4 animations, which have been around for a very long time. If anything though, WebP is a sign that lossy image and video compression are converging, which makes sense. Animated images have a lot more in common with video than stills.

> We could maybe support animated AVIF in the feature, if that ever takes off (which it probably won't since Chrome is removing it).

Do you have a source on that? As far as I know, it's JPEG-XL that they're removing. AVIF is supposed to be the successor to WebP, Google is one of its backers, and it also uses a video codec for image compression.

> Note that you can run `ffmpeg` on GitHub Actions.

Yes, I know. The main concern was that it would consume too much CPU time, but it's probably not too big of a deal anyway. Hosting is a bigger problem, but it's solvable.

Anyway, I might end up fixing this on my end, but it's a bit unfortunate that Tachiyomi won't support this common use-case. Transcoding is unnecessarily lossy, and a waste of CPU cycles.

**Comment by arkon** [comment]
> This is understandable, video codecs are usually part of different libraries/APIs so even with shortcuts, it would still be a non-trivial addition. It's still less complex than a complete video player though.

Exactly. We can probably leverage ExoPlayer to do it, but it's still non-trivial to add it to the paged/webtoon readers properly.

> The Cubari devs have accepted this feature for their frontend, but obviously it's a lot less difficult for them to implement since the browser does all the heavy lifting.

Yes, this is fairly trivial in a browser so long as it ships with the appropriate codecs.

> Do you have a source on that? As far as I know, it's JPEG-XL that they're removing.

Whoops yes, I mixed the two up in my head. Maybe animated AVIFs have a chance. :)

**Comment by 9ary** [comment]
> Exactly. We can probably leverage ExoPlayer to do it, but it's still non-trivial to add it to the paged/webtoon readers properly.

Exoplayer is definitely overkill for this. I was thinking of something along the lines of using Android's media framework which provides all the common codecs. There is basically just one format to support here (h.264 in an MPEG-4 container, with no audio). Get frames from the system decoder, and handle rendering the same way it already works for animated GIF and WebP. I'm more familiar with the internals than the application development side of things but I can't imagine this being terribly complicated. I have too much on my plate to give this a shot myself though.

**Comment by arkon** [comment]
We don't really handle the rendering directly.

We either use:
- For non-animated images: [SubsamplingScaleImageView](https://github.com/tachiyomiorg/subsampling-scale-image-view)
- For animated images: [PhotoView](https://github.com/Baseflow/PhotoView) (if paged reader) or just a plain `AppCompatImageView` (if webtoon reader)

\+ Coil configured with a mix of [our decoder](https://github.com/tachiyomiorg/image-decoder) and the normal Android decoders for some things like the GIF and WebP support.

Coil does supports showing [a single frame](https://coil-kt.github.io/coil/videos/) (which we don't currently have it configured to do) but not full videos (https://github.com/coil-kt/coil/issues/605).

For non-reader contexts (e.g., covers in library/browse/etc.), it's Coil (w/ their Compose UI composable) + our decoder.

The different views are why there's functionality differences between animated and non-animated images in the reader today, which isn't great.

We _could_ just support showing the first frame.

**Comment by 9ary** [comment]
Thanks for the technical details! It does sound like adding support in coil and enabling it here would be the way to go. The different views issue doesn't sound to impact format support, so that's that.

**Comment by arkon** [comment]
I'll leave this open, but I probably won't personally look into adding this anytime soon. If someone else wants to take a stab, feel free.

You'll also need to handle determining the correct format based on MIME type and potentially how to deal with cropping, splitting, and detecting background colors in `ImageUtil` like all other image formats.
